pkg/querier/dispatcher.go

```diff
@@ -0,0 +1,380 @@
+			Labels: mimirpb.FromLabelsToLabelAdapters(s.Labels),
+		})
+	}
+
+	return o.w.Write(ctx, querierpb.EvaluateQueryResponse{
+		Message: &querierpb.EvaluateQueryResponse_SeriesMetadata{
+			SeriesMetadata: &querierpb.EvaluateQueryResponseSeriesMetadata{
+				NodeIndex: o.nodeIndex,
+				Series:    protoSeries,
+			},
+		},
+	})
+}
+
+func (o *evaluationObserver) InstantVectorSeriesDataEvaluated(ctx context.Context, evaluator *streamingpromql.Evaluator, seriesIndex int, seriesData types.InstantVectorSeriesData) error {
+	defer types.PutInstantVectorSeriesData(seriesData, evaluator.MemoryConsumptionTracker)
+
+	// TODO: batch up series to return, rather than sending each series one at a time?
+
+	return o.w.Write(ctx, querierpb.EvaluateQueryResponse{
+		Message: &querierpb.EvaluateQueryResponse_InstantVectorSeriesData{
+			InstantVectorSeriesData: &querierpb.EvaluateQueryResponseInstantVectorSeriesData{
+				NodeIndex:  o.nodeIndex,
+				Floats:     mimirpb.FromFPointsToSamples(seriesData.Floats),
+				Histograms: mimirpb.FromHPointsToHistograms(seriesData.Histograms),
+			},
+		},
+	})
+}
+
+func (o *evaluationObserver) RangeVectorStepSamplesEvaluated(ctx context.Context, evaluator *streamingpromql.Evaluator, seriesIndex int, stepIndex int, stepData *types.RangeVectorStepData) error {
+	// We do not need to return anything to the pool here: the ring buffers in stepData are reused for subsequent steps, or returned to the pool elsewhere if not needed.
+
+	// TODO: batch up series / steps to return, rather than sending each step one at a time?
+
+	floatsHead, floatsTail := stepData.Floats.UnsafePoints()
+	floats, needToReturnFloats, err := combineSlices(floatsHead, floatsTail, types.FPointSlicePool, evaluator.MemoryConsumptionTracker)
+	if err != nil {
+		return err
+	}
+
+	if needToReturnFloats {
+		defer types.FPointSlicePool.Put(&floats, evaluator.MemoryConsumptionTracker)
+	}
+
+	histogramsHead, histogramsTail := stepData.Histograms.UnsafePoints()
+	histograms, needToReturnHistograms, err := combineSlices(histogramsHead, histogramsTail, types.HPointSlicePool, evaluator.MemoryConsumptionTracker)
+	if err != nil {
+		return err
+	}
+
+	if needToReturnHistograms {
+		defer types.HPointSlicePool.Put(&histograms, evaluator.MemoryConsumptionTracker)
+	}
+
+	return o.w.Write(ctx, querierpb.EvaluateQueryResponse{
+		Message: &querierpb.EvaluateQueryResponse_RangeVectorStepData{
+			RangeVectorStepData: &querierpb.EvaluateQueryResponseRangeVectorStepData{
+				NodeIndex:   o.nodeIndex,
+				SeriesIndex: int64(seriesIndex),
+				StepT:       stepData.StepT,
+				RangeStart:  stepData.RangeStart,
+				RangeEnd:    stepData.RangeEnd,
+				Floats:      mimirpb.FromFPointsToSamples(floats),
+				Histograms:  mimirpb.FromHPointsToHistograms(histograms),
+			},
+		},
+	})
+}
+
+func combineSlices[T any](head, tail []T, pool *types.LimitingBucketedPool[[]T, T], memoryConsumptionTracker *limiter.MemoryConsumptionTracker) ([]T, bool, error) {
```

Comment:
Returning this kind of bool flag is a bit strange and not obvious. What if we return some kind of cleanup function instead?

```
If no need to return to the pool
```
```suggestion
func combineSlices[T any](head, tail []T, pool *types.LimitingBucketedPool[[]T, T], memoryConsumptionTracker *limiter.MemoryConsumptionTracker) ([]T, func(), error) {
```

If no need to return thing to pool we can just do 
```
return tail, func(){}, nil // empty function cleanup
```

But if we reach the end of function where we have to return thing to pool we can do something like
```
return combined, func() {
		pool.Put(&combined, memoryConsumptionTracker)
	}, nil
```

Later in the caller we just do the defer of cleanup function
```
floats, cleanup, err := combineSlices(...)
if err != nil {
    return err
}
defer cleanup()
```

Reply:
Good idea, changed in fa01766.